### PR TITLE
MEN-4548: Implement the Query update logic

### PR DESCRIPTION
**Pull request by oleorhagen**
This is based on top of 4547, which may be due to change according to review, but the implementation itself should be pretty freestanding.

### CODE REVIEW

**Review by kacf — CHANGES_REQUESTED** [review]

File: app/updatemanager.go [hunk 1]
```diff
@@ -380,3 +393,116 @@ func removeMatchingElements(m []*UpdateControlMap, cm *UpdateControlMap) (expire
...
+}
+
+// collectPredicates is a utility function to run a 'closure' on all values of
+// the list 'm' matching the predicates in 'predicates...'
+func collectPredicates(m []*UpdateControlMap, f func(*UpdateControlMap), predicates ...func(*UpdateControlMap) bool) {
+	for _, e := range m {
+		// Sentinels
+		for _, predicate := range predicates {
+			if !predicate(e) {
+				return
```
[the remark this answers is not in the document]
Comment (oleorhagen) [reply app/updatemanager.go:503]: Good catch, I just found this myself (I'm using this functionality in `MEN-4547` now too)
Solved it with a `goto`

File: app/updatemanager.go [hunk 2]
```diff
@@ -380,3 +393,116 @@ func removeMatchingElements(m []*UpdateControlMap, cm *UpdateControlMap) (expire
...
 }
+
+// QueryAndUpdate queries the map pool for the correct action to return
+func (c *ControlMapPool) QueryAndUpdate(state string) (action string) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	active, expired := c.flattenMaps()
+	maps := append(active, expired...)
+	sort.Slice(maps, func(i, j int) bool {
+		return maps[i].Priority < maps[j].Priority
```
[the remark this answers is not in the document]
Comment (oleorhagen) [reply app/updatemanager.go:404]: Yes, will do. I really wasn't sure about what should be deemed the higher priority, like in `Unix` it is usually the other way around right. `19` is really _nice_ :), not `-20`

[the remark this answers is not in the document]
Comment (oleorhagen) [reply app/updatemanager.go:404]: A lot of testcases will then need an inversion too :p

File: app/updatemanager.go [hunk 3]
```diff
@@ -305,12 +306,24 @@ func (u *UpdateControlMap) Equal(other *UpdateControlMap) bool {
 	return true
 }
 
+func (u *UpdateControlMap) Action(state string) string {
+	_, ok := u.States[state]
+	if !ok {
+		log.Errorf("Some foobar descriptive error message, cause this should not happen!")
```
[the remark this answers is not in the document]
Comment (oleorhagen) [reply app/updatemanager.go:316]: I'm even tempted to use a panic here...

File: app/updatemanager_test.go [hunk 4]
```diff
@@ -576,3 +576,358 @@ func TestActiveAndExpiredPoolKeys(t *testing.T) {
 		4*time.Second,
 		500*time.Millisecond)
 }
+
+func TestQueryLogic(t *testing.T) {
+	active := 3
+	expired := 0
+	t.Run("Fail exists: Action", func(t *testing.T) {
```
Comment (oleorhagen) [inline app/updatemanager_test.go:583]: All these could probably be written into a nice table test, if you find them hard to decipher

File: app/updatemanager.go [hunk 5]
```diff
@@ -393,3 +406,90 @@ func query(m []*UpdateControlMap, f func(*UpdateControlMap), predicates ...func(
...
+
+// queryActionList returns the next action from the given list,
+// according to this algorithm:
+//
+// 1. If "fail" exists in the list, return "fail".
+// 2. If "pause" exists in the list, return "pause".
+// 2. If "force_continue" exists in the list, return "continue" (<-- note the difference here)
+// 4. "continue" should be ignored, don't return it
+//
+func queryActionList(stateActions []string) string {
```
Comment (oleorhagen) [inline app/updatemanager.go:473]: Not the best name, I just got `query` stuck in my head

File: app/updatemanager.go [hunk 6]
```diff
@@ -393,3 +406,90 @@ func query(m []*UpdateControlMap, f func(*UpdateControlMap), predicates ...func(
...
 	}
 }
+
+// QueryAndUpdate queries the map pool for the correct action to return
+func (c *ControlMapPool) QueryAndUpdate(state string) (action string) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	maps := c.Pool
+	sort.Slice(maps, func(i, j int) bool {
+		return maps[i].Priority > maps[j].Priority
```
Comment (oleorhagen) [inline app/updatemanager.go:416]: Reversed the priorities, as mentioned

File: app/updatemanager.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kacf) [reply app/updatemanager.go]: Hehe, yes, I also thought about that, but they call it "nice value", not priority, so it doesn't really apply. I don't have a very strong preference, but intuitively, in an API it seems natural that a higher number would mean higher priority.

File: app/updatemanager.go (hunk 6 above)
Comment (kacf) [reply app/updatemanager.go:416]: :+1:

**Review by kacf — APPROVED** [review]

**Comment by oleorhagen** [comment]
Waiting for https://github.com/mendersoftware/mender/pull/748 to be merged before continuing

**Comment by oleorhagen** [comment]
Pipe passed previously on the rerun. Only fixed an error log message. Going to merge without rerunning.

**Pull request merged**